index.js

```diff
@@ -12,36 +12,176 @@ var config = require('./lib/config');
-    return tree;
+
+    // Make sure we exlcude files defined in the configuration as well as files from addons
+    var appFiles = new Funnel(tree, {
+      exclude: this._getExcludes()
+    });
+
+    // Instrument JavaScript for code coverage
+    var instrumentedNode = new CoverageInstrumenter(appFiles, {
+      annotation: 'Instrumenting for code coverage',
+      appName: this.parent.pkg.name,
+      appRoot: this.parent.root
+    });
+
+    // Return JavaScript tree with instrumented source
+    return new BroccoliMergeTrees([tree, instrumentedNode], { overwrite: true });
   },
+
+  /**
+   * If coverage is enabled attach coverage middleware to the express server run by ember-cli
+   * @param {Object} startOptions - Express server start options
+   */
+  serverMiddleware: function(startOptions) {
+    this.testemMiddleware(startOptions.app);
+  },
+
+  /**
+   * If coverage is enabled attach coverage middleware to the express server run by testem
+   * @param {Object} app - the Express app instance
+   */
   testemMiddleware: function(app) {
-    if (!this._coverageIsEnabled()) { return; }
+    if (!this._isCoverageEnabled()) { return; }
     attachMiddleware(app, { root: this.project.root, ui: this.ui });
   },
-  serverMiddleware: function(options) {
-    if (!this._coverageIsEnabled()) { return; }
-    attachMiddleware(options.app, { root: this.project.root, ui: this.ui });
+
+  // Custom Methods
+
+  /**
+   * Check if a file exists within the current app/addon
+   * @param {String} relativePath - path to file within current app/addon
+   * @returns {Boolean} whether or not the file exists within the current app/addon
+   */
+  _doesFileExistInCurrentProject: function(relativePath) {
+    relativePath = path.join('app', relativePath);
+
+    try {
+      fs.statSync(relativePath);
+      return true;
+    } catch (err) {
+      return this._doesTemplateFileExist(relativePath);
+    }
   },
-  _coverageIsEnabled: function() {
-    return process.env[this._config().coverageEnvVar];
+
+  /**
+   * Check if a file exists within the dummy app
+   * @param {String} relativePath - path to file within dummy app
+   * @returns {Boolean} whether or not the file exists within the dummy app
+   */
+  _doesFileExistInDummyApp: function(relativePath) {
+    relativePath = path.join('tests', 'dummy', 'app', relativePath);
+
+    try {
+      fs.statSync(relativePath);
+      return true;
+    } catch (err) {
+      return this._doesTemplateFileExist(relativePath);
+    }
+  },
+
+  /**
+   * Check if a template file exists within the current app/addon
+   * Note: Template files are already compiled into JavaScript files so we must
+   * check for the pre-compiled .hbs file
+   * @param {String} relativePath - path to file within current app/addon
+   * @returns {Boolean} whether or not the file exists within the current app/addon
+   */
+  _doesTemplateFileExist: function(relativePath) {
+    relativePath = relativePath.replace('.js', '.hbs');
```

Comment:
We should support any known template extension.  You can get these extensions via:

```js
    var templateExtensions = this.registry.extensionsForType('template');
```

(stolen from https://github.com/ember-cli/ember-cli/blob/d987b4f7dd1e8b2476193b47ac0e5e489cb7f585/lib/models/addon.js#L605)